controllers/pipelinerun_controller.go

```diff
@@ -0,0 +1,215 @@
+	// and only check previously triggered builds when the name matches
+	var triggeredBuilds = []filter.TriggeredBuild{}
+	if filter.PipelineRunAnnotatedNameMatchesObject(&pipelineRun) {
+		// extracting existing triggered-builds from the annotation, information needed to detect if
+		// the BuildRuns have already beeing issued for the PipelineRun
+		triggeredBuilds, err = filter.PipelineRunExtractTriggeredBuildsSlice(&pipelineRun)
+		if err != nil {
+			logger.V(0).Error(err, "parsing triggered-builds annotation")
+			// in case of errors an empty slice takes place, may incur the side effect of issuing
+			// duplicated BuildRuns
+			triggeredBuilds = []filter.TriggeredBuild{}
+		}
+
+		// filtering out the instances that have already been processed, the annotation extracted
+		// shows  which build names and the objectRef employed
+		if filter.TriggereBuildsContainsObjectRef(triggeredBuilds, buildNames, objectRef) {
+			logger.V(0).Info("BuildRuns for PipelineRun have already been issued!")
+			return ctrl.Result{}, nil
+		}
+	} else {
+		logger.V(0).Info("PipelineRun annotated name does not match current object!")
+	}
+	logger.V(0).Info("PipelineRun previously triggered builds", "triggered-builds", triggeredBuilds)
+
+	// firing the BuildRun instances for the informed Builds
+	buildRunsIssued, err := r.issueBuildRunsForPipelineRun(ctx, &pipelineRun, buildNames)
+	if err != nil {
+		logger.V(0).Error(err, "trying to issue BuildRun instances")
+		return ctrl.Result{}, err
+	}
+	logger.V(0).Info("BuildRuns issued", "buildruns", buildRunsIssued)
+
+	// updating annotation appending the current state which triggered BuildRuns instances, this
+	// annotation is later on checked to skip the conditions that already triggered builds
+	if err = filter.PipelineRunAppendTriggeredBuildsAnnotation(
+		&pipelineRun,
+		triggeredBuilds,
+		buildNames,
+		objectRef,
+	); err != nil {
+		logger.V(0).Error(err, "trying to updated triggered-builds annotation")
+		return ctrl.Result{}, err
+	}
+
+	// updating label registering all BuildRuns issued
+	filter.AppendIssuedBuildRunsLabel(&pipelineRun, buildRunsIssued)
+	// annotating object's current name
+	filter.PipelineRunAnnotateName(&pipelineRun)
+
+	// patching the PipelineRun to reflect labels and annotations needed on the object
+	if err = r.Client.Patch(ctx, &pipelineRun, client.Merge); err != nil {
+		logger.V(0).Error(err, "trying to update PipelineRun metadata")
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager uses the manager to watch over PipelineRuns.
+func (r *PipelineRunReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if r.Clock == nil {
+		r.Clock = realClock{}
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&tknv1beta1.PipelineRun{}).
+		WithEventFilter(predicate.NewPredicateFuncs(filter.EventFilterPredicate)).
+		WithEventFilter(predicate.Funcs{
+			DeleteFunc: func(e event.DeleteEvent) bool {
```

Comment:
Should we filter here also the UpdateFunc ? I think we only are interested in updates where certain fields of the status are updated: startTime, conditions[type=Succeeded].Status.

Reply:
We already filter the objects, it only allows the objects with the interesting fields populated. Please check the `EventFilterPredicate`.